security/elastic.py

```diff
@@ -0,0 +1,152 @@
+#    under the License.
+
+import datetime
+import json
+import logging
+
+import elasticsearch
+from elasticsearch import helpers
+
+from security import base
+
+INDEX = "ms_security_"
+ELASTIC_DATETIME_FORMAT = "yyyy-MM-dd HH:mm:ss Z"
+PY_DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S +00"
+
+
+def elastic_to_issue(dict_):
+    for attr in ("discovered_at", "confirmed_at", "resolved_at"):
+        value = dict_.get(attr)
+        if value:
+            dict_[attr] = datetime.datetime.strptime(value, PY_DATETIME_FORMAT)
+    return dict_
+
+
+def issue_to_elastic(issue):
+    dict_ = {
+        "type": issue.type,
+        "description": issue.description,
+        "subject": issue.subject,
+        "discovered_at": issue.discovered_at.strftime(PY_DATETIME_FORMAT),
+        "confirmed_at": issue.confirmed_at.strftime(PY_DATETIME_FORMAT),
+    }
+    if issue.resolved_at:
+        dict_["resolved_at"] = issue.resolved_at.strftime(PY_DATETIME_FORMAT)
+    return dict_
+
+
+class StorageException(Exception):
+    pass
+
+
+class Backend(object):
+    """Represent elasticsearch backend.
+
+    usage:
+        b = elastic.Backend(urls=["url1", "url2"])
+        issues = b.get_issues("Region1", "IssueType1,IssueType2")
+        b.begin()
+        b.store_issue(Issue(...))
+        b.commit("region1")
+    """
+
+    def __init__(self, urls, **kwargs):
+        self.index = INDEX
+        self.es = elasticsearch.Elasticsearch(urls)
+        self.body = []
+
+    def get_issues(self, region, types=None, discovered_days=None):
+        query = {
+            "query": {
+                "bool": {
+                    "must_not": {
+                        "exists": {
+                            "field": "resolved_at",
+                        },
+                    },
+                },
+            }
+        }
+        if discovered_days:
+            query["query"]["bool"]["must"] = {
+                "range": {
+                    "discovered_at": {"gte": "now-%dd/h" % discovered_days}
+                }
+            }
+        issues = []
+        try:
+            for i in helpers.scan(self.es, index=self.index + region,
+                                  doc_type=types, query=query, scroll="1m"):
+                kwargs = i["_source"]
+                logging.debug("Loaded issue %s" % kwargs)
+                issues.append(base.Issue(i["_type"],
+                              **elastic_to_issue(kwargs)))
+            return issues
+        except elasticsearch.NotFoundError:
+            return []
+
+    def begin(self):
+        if self.body:
+            logging.warning("Unsaved data %s", self.body)
+        self.body = ""
+
+    def commit(self, region):
+        if not region:
+            raise StorageException("Missing region")
+        try:
+            self.es.indices.create(index=self.index + region, body={
+                "mappings": {
+                    "_default_": {
+                        "properties": {
```

Comment:
it's very important to specify all fields in Elastic and use Elastic in strict mode, if you add any text field in this index it won't be indexed properly.